From 19db6eb73a8e88002b358f5d2ccbf96eb34987ab Mon Sep 17 00:00:00 2001
From: fujiwarat <takao.fujiwara1@gmail.com>
Date: Thu, 11 Nov 2010 18:09:52 +0900
Subject: [PATCH] Fix 32bit big endian sequences for X11 XNClientWindow & XNFocusWindow

---
 client/x11/main.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/client/x11/main.c b/client/x11/main.c
index c91a6d7..be9cb0e 100644
--- a/client/x11/main.c
+++ b/client/x11/main.c
@@ -277,10 +277,10 @@ _xim_store_ic_values (X11IC *x11ic, IMChangeICStruct *call_data)
             x11ic->input_style = *(gint32 *) ic_attr->value;
         }
         else if (g_strcmp0 (XNClientWindow, ic_attr->name) == 0) {
-            x11ic->client_window =  *(Window *) call_data->ic_attr[i].value;
+            x11ic->client_window =  (Window)(*(CARD32 *) call_data->ic_attr[i].value);
         }
         else if (g_strcmp0 (XNFocusWindow, ic_attr->name) == 0) {
-            x11ic->focus_window =  *(Window *) call_data->ic_attr[i].value;
+            x11ic->focus_window =  (Window)(*(CARD32 *) call_data->ic_attr[i].value);
         }
         else {
             LOG (1, "Unknown ic attribute: %s", ic_attr->name);
-- 
1.7.2.1

