--- ibus-1.3.4/client/gtk2/ibusimcontext.c.orig	2014-07-22 19:43:01.219540764 +0900
+++ ibus-1.3.4/client/gtk2/ibusimcontext.c	2015-10-13 16:02:24.011008653 +0900
@@ -947,8 +947,16 @@ _ibus_context_destroy_cb (IBusInputConte
     g_free (ibusimcontext->preedit_string);
     ibusimcontext->preedit_string = NULL;
 
+    if (G_OBJECT (ibusimcontext)->ref_count == 0)
+        return;
+    /* rhbz:1245288 gtk_im_multicontext_get_slave() unref ibusimcontext
+     * if gtk-im-module is changed.
+     */
+    g_object_ref (ibusimcontext);
     g_signal_emit (ibusimcontext, _signal_preedit_changed_id, 0);
     g_signal_emit (ibusimcontext, _signal_preedit_end_id, 0);
+    if (G_OBJECT (ibusimcontext)->ref_count > 1)
+        g_object_unref (ibusimcontext);
 }
 
 static void
